### PR TITLE
fix: refactor the LoadPolicy in SyncedEnforcer, reduce the lock range

### LINKED ISSUE #1108: [Feature] Cut down the lock range in SyncedEnforcer
In SyncedEnforcer, LoadPolicy is a heavy time cost function if there are a lots of policies.

https://github.com/casbin/casbin/blob/f58b2c99df610c757ddbda947f985d7f566709e1/enforcer_synced.go#L117-L121

It can use the lock only when access the critical section.

### PR DESCRIPTION
**Pull request by deagon**
fix #1108

### CODE REVIEW

**Comment by hsluoyz** [comment]
@deagon what does this PR do?

**Comment by deagon** [comment]
> @deagon what does this PR do?

@hsluoyz fix https://github.com/casbin/casbin/issues/1108

**Comment by hsluoyz** [comment]
@JalinWang plz review

/cc @tangyang9464

**Comment by JalinWang** [comment]
The pr is technically ok. However, when `LoadPolicy` is called, it indicates changes in policies. If the function starts without locking, there might be other enforcement requests using out-of-date policies.

**Comment by deagon** [comment]
> The pr is technically ok. However, when `LoadPolicy` is called, it indicates changes in policies. If the function starts without locking, there might be other enforcement requests using out-of-date policies.

That's right. However, `LoadPolicy` costs almost 3 seconds in my production environment. The permission read requests would be blocked 3 seconds If it locks, which is unacceptable for us.
In this case, I think out-of-date policies might be ok.

File: enforcer_synced.go [hunk 1]
```diff
@@ -115,9 +117,42 @@ func (e *SyncedEnforcer) ClearPolicy() {
 
 // LoadPolicy reloads the policy from file/database.
 func (e *SyncedEnforcer) LoadPolicy() error {
```
Comment (hsluoyz) [inline enforcer_synced.go:119]: Define a new function named `LoadPolicyFast()` to hold your code

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by deagon** [comment]
@hsluoyz  Please take a review, thanks.

**Review by hsluoyz — APPROVED** [review]

**Comment by hsluoyz** [comment]
@deagon hi, this is impressive PR! Plz join the QQ group: 540163681

**Comment by pinkyo** [comment]
thanks a lot, it help us solved a global performance issue. we got 400k policy data in db, and a reload operation cost more than 10s.

**Pull request merged**